File: MCRunner/Strategy/StrategyManager.cs
using Bar = MCRunner.Instruments.Bar;
using MCRunner.Orders;
using MCRunner.Positions;
using PowerLanguage;
using System.Collections.Immutable;
using System;

namespace MCRunner.Strategy
{
    public class StrategyManager
    {
        public event Action<OrderInfo> OrderValidated;

        public event Action<OrderInfo> OrderCanceled;

        private ImmutableHashSet<OrderInfo> previousSubmittedOrders =
            ImmutableHashSet<OrderInfo>.Empty;
        private ImmutableHashSet<OrderInfo> untriggeredOrders =
            ImmutableHashSet<OrderInfo>.Empty;
        private ImmutableHashSet<OrderInfo> validatedOrders =
            ImmutableHashSet<OrderInfo>.Empty;
        private readonly IPosition longPosition = new LongPosition();
        private readonly IPosition shortPosition = new ShortPosition();
        private readonly StrategyPerformance strategyInfo = new StrategyPerformance();

        private int Size
        {
            get
            {
                return (int)(longPosition.Size - shortPosition.Size);
            }
        }

        public IStrategyPerformance StrategyInfo
        {
            get
            {
                return strategyInfo;
            }
        }

        public StrategyManager(IOrderManaged orderCreator)
        {
            orderCreator.OrderSent += OnOrderSent;
        }

        public void UpdateMarketPositionAtBroker(double positionChange)
        {
            strategyInfo.UpdateMarketPositionAtBroker(positionChange);
        }

        public void TriggerOrders(Bar bar)
        {
            foreach (var order in untriggeredOrders)
            {
                if (order.Order is IOrderPriced)
                {
                    var price = order.ConditionPrice;
                    if (price <= 0)
                    {
                        price = order.Price;
                    }

                    switch (order.OrderAction)
                    {
                        case EOrderAction.Buy:
                        case EOrderAction.BuyToCover:
                            if (price >= bar.Low)
                            {
                                OnOrderTriggered(order);
                            }
                            break;

                        case EOrderAction.Sell:
                        case EOrderAction.SellShort:
                            if (price <= bar.High)
                            {
                                OnOrderTriggered(order);
                            }
                            break;

                        default:
                            throw new ArgumentException("Unknown OrderAction");
                    }
                }
                else
                {
                    throw new ArgumentException("Unknown order type");
                }
            }

            foreach (var order in previousSubmittedOrders)
            {
                if (!untriggeredOrders.Contains(order))
                {
                    OrderCanceled.SafeTrigger(order);
                }
            }

            previousSubmittedOrders = untriggeredOrders;
            untriggeredOrders = untriggeredOrders.Clear();
        }

        private void OnOrderSent(OrderInfo order)
        {
            ValidateOrder(order);

            if (!validatedOrders.Contains(order))
            {
                validatedOrders = validatedOrders.Add(order);
                OrderValidated.SafeTrigger(order);
            }

            CheckIfOrderTriggered(order);
        }

        private void OnOrderTriggered(OrderInfo order)
        {
            switch (order.OrderAction)
            {
                case EOrderAction.Buy:
                case EOrderAction.Sell:
                    longPosition.UpdatePosition(order);
                    break;

                case EOrderAction.SellShort:
                case EOrderAction.BuyToCover:
                    shortPosition.UpdatePosition(order);
                    break;

                default:
                    throw new ArgumentException("Invalid OrderAction");
            }

            strategyInfo.MarketPosition = Size;

            validatedOrders = validatedOrders.Remove(order);
            untriggeredOrders = untriggeredOrders.Remove(order);
            previousSubmittedOrders = previousSubmittedOrders.Remove(order);
        }

        private void ValidateOrder(OrderInfo order)
        {
            switch (order.OrderAction)
            {
                case EOrderAction.Buy:
                case EOrderAction.Sell:
                    longPosition.ValidateOrder(order);
                    break;

                case EOrderAction.SellShort:
                case EOrderAction.BuyToCover:
                    shortPosition.ValidateOrder(order);
                    break;

                default:
                    throw new InvalidOrderException("Unknown OrderAction");
            }
        }

        private void CheckIfOrderTriggered(OrderInfo order)
        {
            if (order.Order is IOrderMarket)
            {
                OnOrderTriggered(order);
            }
            else
            {
                untriggeredOrders = untriggeredOrders.Add(order);
            }
        }
    }
}

File: python/mcrunner/strategy/strategy_manager.py
from typing import List, Callable
from ..orders import (
    OrderCreator,
    OrderInfo,
    EOrderAction,
    OrderCategory,
)
from ..positions import LongPosition, ShortPosition
from .strategy_performance import StrategyPerformance
from .trade import ExecutedTrade

class StrategyManager:
    def __init__(self, order_creator: OrderCreator):
        self._long = LongPosition()
        self._short = ShortPosition()
        self.StrategyInfo = StrategyPerformance()
        self.ExecutedOrders: List[ExecutedTrade] = []
        self._untriggered: List[OrderInfo] = []
        self._prev_submitted: List[OrderInfo] = []
        self.OrderValidated: List[Callable[[OrderInfo], None]] = []
        self.OrderCanceled: List[Callable[[OrderInfo], None]] = []
        order_creator.order_sent = self.on_order_sent

    def on_order_sent(self, order: OrderInfo):
        self.validate_order(order)
        if order not in self._untriggered:
            self._untriggered.append(order)
            for cb in self.OrderValidated:
                cb(order)

    def validate_order(self, order: OrderInfo):
        if order.OrderAction in (EOrderAction.Buy, EOrderAction.Sell):
            self._long.validate_order(order)
        else:
            self._short.validate_order(order)

    def trigger_orders(self, bar, index: int):
        triggered = []
        for order in list(self._untriggered):
            if order.Order.Category == OrderCategory.Market:
                triggered.append(order)
            elif order.Order.Category in (OrderCategory.Limit, OrderCategory.Stop, OrderCategory.StopLimit):
                price = order.ConditionPrice or order.Price
                if order.OrderAction in (EOrderAction.Buy, EOrderAction.BuyToCover):
                    if price >= getattr(bar, 'Low', 0):
                        triggered.append(order)
                else:
                    if price <= getattr(bar, 'High', 0):
                        triggered.append(order)
        for order in triggered:
            self.on_order_triggered(order, bar, index)

        for order in list(self._prev_submitted):
            if order not in self._untriggered:
                for cb in self.OrderCanceled:
                    cb(order)
        self._prev_submitted = self._untriggered
        self._untriggered = []

    def on_order_triggered(self, order: OrderInfo, bar, index: int):
        if order.Price <= 0:
            order.Price = getattr(bar, 'Close', 0)
        self.ExecutedOrders.append(ExecutedTrade(index, order.OrderAction, order.Price, order.Size))
        if order.OrderAction in (EOrderAction.Buy, EOrderAction.Sell):
            self._long.update_position(order)
        else:
            self._short.update_position(order)
        self.StrategyInfo.MarketPosition = int(self._long.Size - self._short.Size)
        if order in self._untriggered:
            self._untriggered.remove(order)
        if order in self._prev_submitted:
            self._prev_submitted.remove(order)
